### PR TITLE
Enable SQLite data to be read from cloud-based sources

### PR DESCRIPTION
**Pull request by d33bs**
# Description

This PR addresses #43 by enabling SQLite data to be read from cloud-based locations. This work involves caching cloud-based SQLite data locally so they may be queried locally. Please see the documentation updates for more details on the exact mechanism and examples.

Note, while addressing changes, I also made minor updates which are tangential but help with the performance or understanding of the changes:
- Updated documentation to exclude the use of `merge` and instead prefer `join` (leaning towards SQL-based rather than Pandas-based data linking language).
- Marked a single dask test as skipped by default. This test is memory intensive and I believe is the cause of Github Actions testing resource issues. Leaving the test will allow for further testing of this capability if desired and hopefully avoid challenges with automated testing in Github. In skipping the test I also hope to make more room for the additional testing performed to validate the new functionality added by this PR.

## What is the nature of your change?

- [x] Bug fix (fixes an issue).
- [ ] Enhancement (adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] This change requires a documentation update.

# Checklist

Please ensure that all boxes are checked before indicating that a pull request is ready for review.

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have deleted all non-relevant text in this pull request template.

### CODE REVIEW

File: cytotable/sources.py
Comment (gwaybio) [inline cytotable/sources.py:55]:
```suggestion
            Either a directory path to seek filepaths within or a path directly to a file.
```

Consider also specifying something to the extend that the Path can be local or on the cloud.

Comment (gwaybio) [reply cytotable/sources.py:55]: suggestion just to boost documentation clarity

File: cytotable/utils.py
Comment (gwaybio) [inline cytotable/utils.py:111]: Do you have a sense of how long this will take? I presume this will be slow

**Review by gwaybio — APPROVED** [review]
LGTM!

We should wait for @falquaddoomi 's review to handle more of the technical elements

File: cytotable/utils.py
Comment (falquaddoomi) [inline cytotable/utils.py:116]: From the docs, seems that just referencing `.fspath` is enough to cause the remote file to be cached, which I imagine would make it unnecessary to call `read_bytes()` on it: https://cloudpathlib.drivendata.org/stable/caching/#accessing-the-cached-version-directly-read-only.

Also, I could imagine `clouthpathlib` being extended with additional providers in the future. If that seems like something you'd want, you might defer to the library to determine if it supports the given URL, e.g.:

```python
candidate_path = AnyPath(path)
  
if candidate_path.is_file() and candidate_path.suffix.lower() == ".sqlite":
  try:
    path = CloudPath(path).fspath
  except clouthpathlib.exceptions.CloudPathException:
    # you might consider logging here, up to you
    pass
    
return pathlib.Path(path)
```

**Review by falquaddoomi — COMMENTED** [review]
Nice work; I left a few comments.

File: cytotable/utils.py
Comment (d33bs) [reply cytotable/utils.py:111]: This will vary based on where CytoTable is running (network access, hardware data storage type), the cloud provider (AWS, GCP, etc.), and how the bucket is configured (geolocation, throttling, etc). For AWS S3 locations [`cloudpathlib` leverages `boto3` via the `S3Client`](https://github.com/drivendataorg/cloudpathlib/blob/master/cloudpathlib/s3/s3client.py#L153) to perform downloads.

Wondering aloud: would it be a good idea to provide a caveat about larger cloud-based SQLite data (that performance may vary)?

File: cytotable/sources.py
Comment (d33bs) [reply cytotable/sources.py]: So much better, thanks!

File: cytotable/utils.py
Comment (d33bs) [reply cytotable/utils.py:116]: Thanks for this, great thoughts! I wrestled with both of these a bit and welcome the chance to improve.
- `fspath` vs `read_bytes`: great point, I'll get this changed. I don't feel strongly about it but wondering aloud if `read_bytes` is less presumptive about the underlying operations (perhaps a cloudpath could be understood to be a type of "filesystem" in some circumstances).
- additional providers: One reason for the explicit checks was that I couldn't get [`CloudPath.is_valid_cloudpath`](https://cloudpathlib.drivendata.org/v0.6/api-reference/cloudpath/#cloudpathlib.cloudpath.CloudPath.is_valid_cloudpath) to work in expected ways. I like the alternative of a try/except here better for the reasons you mentioned. I'll these a try in an update.

File: cytotable/sources.py
Comment (d33bs) [reply cytotable/sources.py:220]: Thanks! This was accidental during the course of development, mostly with regards to how cloud directories vs cloud files could be handled within `cytotable`. I like the `and` better for readability and will move back to this.

File: cytotable/sources.py
Comment (d33bs) [reply cytotable/sources.py:85]: Thanks, this is great! I'll move over to this new format as you suggested; the edge case of an empty list is unintentional (maybe at some point property-based / hypothesis testing in this project could capture things like this!).

**Review by falquaddoomi — APPROVED** [review]

**Pull request merged**
